File: app/services/workbench/group_candidates.py
"""rows → CandidateGroups (L2) — mirror of client group-candidates.ts."""

from __future__ import annotations

from app.schemas.workbench_schema import CandidateGroups, ReplacementPlanV1, WorkbenchCandidate
from app.services.workbench.candidate_rank import candidate_sort_key, relation_index
from app.services.workbench.candidate_reason import candidate_reasons


def group_candidates(plan: ReplacementPlanV1, rows: list[dict], pool) -> CandidateGroups:
    direct = relation_index(pool.syns) if pool else {}
    semantic = relation_index(pool.semantic) if pool else {}
    groups: dict[str, list[WorkbenchCandidate]] = {
        "direct_syn": [],
        "semantic_related": [],
        "sound_only": [],
    }
    for row_rank, row in enumerate(rows):
        literal = str(row.get("char") or row.get("literal") or "")
        if not literal:
            continue
        if literal in direct:
            group = "direct_syn"
            source_rank, source = direct[literal]
        elif literal in semantic:
            group = "semantic_related"
            source_rank, source = semantic[literal]
        else:
            group = "sound_only"
            source_rank, source = row_rank, None

        if plan.semantic_intent == "direct_only" and group != "direct_syn":
            continue
        if plan.semantic_intent == "off":
            group, source_rank, source = "sound_only", row_rank, None

        code = str(row.get("code") or "")
        groups[group].append(WorkbenchCandidate(
            literal=literal,
            jyutping=str(row.get("jyutping") or ""),
            code=code,
            group=group,
            reasons=candidate_reasons(plan, code, group, relation_source=source),
            source_rank=source_rank,
        ))
    for values in groups.values():
        values.sort(key=candidate_sort_key)
    return CandidateGroups(**groups)


def group_literals(groups: CandidateGroups) -> dict[str, list[str]]:
    return {
        "direct_syn": [c.literal for c in groups.direct_syn],
        "semantic_related": [c.literal for c in groups.semantic_related],
        "sound_only": [c.literal for c in groups.sound_only],
    }


def candidate_count_for_pool(plan: ReplacementPlanV1, rows: list[dict], pool) -> int:
    if plan.semantic_intent != "direct_only":
        return len(rows)
    direct = relation_index(pool.syns) if pool else {}
    return sum(1 for row in rows if str(row.get("char") or "") in direct)


__all__ = ["candidate_count_for_pool", "group_candidates", "group_literals"]

File: client/src/workbench/group-candidates.ts
import type { WordRow } from '../db/position-match/word-row.ts';
import { candidateReasons } from './candidate-reason.ts';
import { compareSoundOnlyCandidates, relationIndex } from './candidate-rank.ts';
import type {
  CandidateGroup,
  CandidateGroups,
  ReplacementPlanV1,
  WorkbenchCandidate,
} from './contracts.ts';

/** Minimal pool surface for grouping (full RelationPoolSnapshot is fine). */
export type GroupPoolInput = {
  syns?: Array<{ char: string; source?: string }>;
  semantic?: Array<{ char: string; source?: string }>;
} | null;

function compareRelationCandidates(a: WorkbenchCandidate, b: WorkbenchCandidate): number {
  return a.sourceRank - b.sourceRank
    || a.literal.localeCompare(b.literal)
    || a.jyutping.localeCompare(b.jyutping);
}

/** rows + optional relation pool → CandidateGroups（L2）。 */
export function groupCandidates(
  plan: ReplacementPlanV1,
  rows: WordRow[],
  pool: GroupPoolInput,
): CandidateGroups {
  const direct = relationIndex(pool?.syns ?? []);
  const semantic = relationIndex(pool?.semantic ?? []);
  const groups: CandidateGroups = { direct_syn: [], semantic_related: [], sound_only: [] };
  rows.forEach((row, rowRank) => {
    const literal = String(row.char ?? '');
    if (!literal) return;
    let group: CandidateGroup = direct.has(literal) ? 'direct_syn'
      : semantic.has(literal) ? 'semantic_related' : 'sound_only';
    const relation = group === 'direct_syn' ? direct.get(literal) : semantic.get(literal);
    if (plan.semanticIntent === 'direct_only' && group !== 'direct_syn') return;
    if (plan.semanticIntent === 'off') group = 'sound_only';
    const code = String(row.code ?? '');
    const candidate: WorkbenchCandidate = {
      literal,
      jyutping: String(row.jyutping ?? ''),
      code,
      group,
      reasons: candidateReasons(plan, code, group, relation?.source),
      sourceRank: plan.semanticIntent === 'off' ? rowRank : relation?.rank ?? rowRank,
    };
    groups[group].push(candidate);
  });
  groups.direct_syn.sort(compareRelationCandidates);
  groups.semantic_related.sort(compareRelationCandidates);
  groups.sound_only.sort(compareSoundOnlyCandidates);
  return groups;
}

/** Literals only per group — for L2 fixture parity without reason noise. */
export function groupLiterals(groups: CandidateGroups): Record<CandidateGroup, string[]> {
  return {
    direct_syn: groups.direct_syn.map((c) => c.literal),
    semantic_related: groups.semantic_related.map((c) => c.literal),
    sound_only: groups.sound_only.map((c) => c.literal),
  };
}
